File: 큰수만들기.py
from collections import deque
def solution(number, k):
    answer = ''
    n = len(number)-k
    stack = deque() 
    arr = list(map(int, number))
    arr = deque(arr)
    while len(arr)>0:
        if len(stack)==0:
            stack.append(arr.popleft())
            continue
        else:
            top = stack[-1]
            push_value = arr.popleft()
            if top < push_value:
                while top<push_value and k>0:
                    stack.pop()
                    k-=1
                    if len(stack)>0: top = stack[-1]
                    else: break
                stack.append(push_value)
            else: stack.append(push_value)
    if len(stack)>n:
        while len(stack) != n:
            stack.pop()
    if stack[0] == 0:
        answer = '0'
    else:
        for i in stack:
            answer += str(i)
    return answer

File: 큰수만들기.java
import java.util.*;
class Solution {
    public String solution(String number, int k) {
        String answer = "";
        ArrayList<Integer> arr = new ArrayList<>();
        for(int i=0;i<number.length();i++){
            arr.add(Integer.parseInt(number.substring(i,i+1)));
        }
        Deque<Integer> temp = new ArrayDeque<>();
        
        for(int i=0;i<arr.size();i++){
            while(k>0 && temp.size()>0 && temp.peekLast()<arr.get(i)){
                k--;
                temp.pollLast();
            }
            temp.add(arr.get(i));
        }
     
        while(temp.size()>0){
            answer += temp.pollFirst();
        }
        if(k>0) answer = answer.substring(0, k+1);
        return answer;
    }
}
